### PR TITLE
[LinearAlgebra] Add functions ZTOOL-116

### PR DESCRIPTION
**Pull request by ArianJM**
Jira: https://zybooks.atlassian.net/browse/ZTOOL-116

Adds the new functions.

### CODE REVIEW

**Comment by ArianJM** [comment]
@AlexEdgcomb ready for you.

File: src/lib/linear_algebra/__init__.py [hunk 1]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
         latex_output += ' \\\\ '
     latex_output += '\\end{align}'
     return latex_output
+
+def transpose_matrix(matrix):
+    '''
+    Returns the transpose of a matrix.
+    '''
+    return map(list,zip(*matrix))
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: I think there should be a space after the comma

File: src/lib/linear_algebra/__init__.py [hunk 2]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
+
+def get_minor_matrix(matrix, row_in_matrix, col_in_matrix):
+    '''
+    Returns the submatrix of m used to calculate the minor corresponding to m[i][j].
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: What is `m`? I think it's `matrix`

File: src/lib/linear_algebra/__init__.py [hunk 3]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
+    Example: Let m = [ [1, 2, 3], [4, 5, 6], [7, 8, 9] ], i = 0, j = 0
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: Again, I think `m` is `matrix`

File: src/lib/linear_algebra/__init__.py [hunk 4]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
     latex_output += '\\end{align}'
     return latex_output
+
+def transpose_matrix(matrix):
+    '''
+    Returns the transpose of a matrix.
+    '''
+    return map(list,zip(*matrix))
+
+def get_minor_matrix(matrix, row_in_matrix, col_in_matrix):
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: Should we full write-out `column`? Ex: `column_in_matrix` ?

File: src/lib/linear_algebra/__init__.py [hunk 5]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
+    '''
+    if len(matrix) != len(matrix[0]):
+        raise ValueError('Inverse of a non-square matrix is undefined')
+
+    determinant = get_determinant(matrix)
+    if determinant == 0:
+        raise ValueError('Matrix is singular, inverse does not exist')
+
+    size_of_matrix = len(matrix)
+    adjugate_matrix = [ [get_determinant(get_minor_matrix(matrix, col, row)) * (-1) ** (col + row) for col in range(size_of_matrix)] for row in range(size_of_matrix) ]
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: I think `get_determinant(get_minor_matrix(matrix, col, row)) * (-1) ** (col + row)` could use one more set of parens.

Maybe: `get_determinant(get_minor_matrix(matrix, col, row)) * ((-1) ** (col + row))` ?

Comment (AlexEdgcomb) [reply src/lib/linear_algebra/__init__.py]: Also, there needs to be a space after `[`. Ex: `[ [ get_determinant(`

Comment (AlexEdgcomb) [reply src/lib/linear_algebra/__init__.py]: I think it'd also be good to split this into a few lines. It's so long!

File: src/lib/linear_algebra/__init__.py [hunk 6]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
+    inverse_matrix = [ [adjugate_matrix[row][col] / float(determinant) for col in range(size_of_matrix)] for row in range(size_of_matrix) ]
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: Similar comments as above, as well as, spelling out `column`

Maybe:
```
inverse_matrix = [
    [ adjugate_matrix[row][column] / float(determinant) for column in range(size_of_matrix) ]
    for row in range(size_of_matrix)
]
```

File: src/lib/linear_algebra/__init__.py [hunk 7]
```diff
@@ -233,3 +233,71 @@ def generate_equations_from_matrix(matrix):
...
+def matrix_to_ints(matrix):
+    '''
+    Converts all of the entries in a matrix to integers.
+    This is useful so that student responses in matrix questions are checked precisely.
+    Ex:
+    matrix = [ [1.0, 2.0], [-1.0, 0.0] ]
+    int_matrix = matrix_to_ints(matrix)
+    int_matrix is [ [1, 2], [-1, 0] ]
+    '''
+    int_matrix = [ [int(matrix[i][j]) for j in range(len(matrix[0]))] for i in range(len(matrix)) ]
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: I don't think a variable is needed; we can just return the created list. Also, I don't think `range(len(...))` is clearest. 

How about:
```
return [
    [ int(element) for element in row_elements ]
    for row_elements in matrix
]
```

**Comment by ronsiu** [comment]
@AlexEdgcomb Comments addressed.

File: src/lib/linear_algebra/__init__.py [hunk 8]
```diff
@@ -233,3 +233,86 @@ def generate_equations_from_matrix(matrix):
...
+    return map(list, zip(*matrix))
+
+def get_minor_matrix(matrix, row_in_matrix, column_in_matrix):
+    '''
+    Returns the submatrix of matrix used to calculate the minor corresponding to matrix[row_in_matrix][column_in_matrix].
+    Example: Let matrix = [ [1, 2, 3], [4, 5, 6], [7, 8, 9] ], row_in_matrix = 0, column_in_matrix = 0
+    Function returns: [ [5, 6], [8, 9] ]
+    More info: https://en.wikipedia.org/wiki/Minor_(linear_algebra)
+    '''
+    return [ row[:column_in_matrix] + row[column_in_matrix + 1:] for row in (matrix[:row_in_matrix] + matrix[row_in_matrix + 1:]) ]
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: This line is very long

File: src/lib/linear_algebra/__init__.py [hunk 9]
```diff
@@ -233,3 +233,86 @@ def generate_equations_from_matrix(matrix):
...
+    size_of_matrix = len(matrix)
+
+    # Base case: the determinant of a 1 x 1 matrix is the single entry of the matrix
+    if size_of_matrix == 1:
+        return matrix[0][0]
+
+    # Recursive case: the determinant of an n x n matrix is the sum of cofactors along the top row
+    determinant = 0
+    for cofactor in range(size_of_matrix):
+        determinant += ((-1) ** cofactor ) * matrix[0][cofactor] * get_determinant(get_minor_matrix(matrix, 0, cofactor))
```
Comment (AlexEdgcomb) [inline src/lib/linear_algebra/__init__.py]: This line is very long

**Comment by ronsiu** [comment]
@AlexEdgcomb Comment addressed

**Comment by AlexEdgcomb** [comment]
@ronsiu , code looks great. Please verify functionality of changed functions, so Arian knows they work :)

**Comment by ronsiu** [comment]
@AlexEdgcomb Confirmed functionality is ok

**Comment by AlexEdgcomb** [comment]
@ArianJM , I'll merge if you agree :)

**Comment by ArianJM** [comment]
@AlexEdgcomb yes, let's merge.

**Pull request merged**